csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll

```diff
@@ -0,0 +1,181 @@
+private import codeql.util.Unit
+private import codeql.util.FilePath
+private import semmle.code.csharp.frameworks.microsoft.AspNetCore
+
+/** A call to the `View` method */
+private class ViewCall extends MethodCall {
+  ViewCall() { this.getTarget().hasQualifiedName("Microsoft.AspNetCore.Mvc", "Controller", "View") }
+
+  /** Gets the `name` argument to this call, if any. */
+  string getNameArgument() {
+    exists(StringLiteral lit, int i | i in [0 .. 1] |
+      this.getTarget().getParameter(i).getType() instanceof StringType and
+      DataFlow::localExprFlow(lit, this.getArgument(i)) and
+      result = lit.getValue()
+    )
+  }
+
+  /** Gets the `model` argument to this call, if any. */
+  Expr getModelArgument() {
+    exists(int i | i in [0 .. 1] |
+      this.getTarget().getParameter(i).getType() instanceof ObjectType and
+      result = this.getArgument(i)
+    )
+  }
+
+  /** Gets the MVC action method that this call is made from, if any. */
+  Method getActionMethod() {
+    result = this.getEnclosingCallable() and
+    result = this.getController().getAnActionMethod()
+  }
+
+  /**
+   * Gets the action name that this call refers to, if any.
+   * This is either the name argument, or the name of the action method calling this if there is no name argument.
+   */
+  string getActionName() {
+    result = this.getNameArgument()
+    or
+    not exists(this.getNameArgument()) and
+    result = this.getActionMethod().getName()
+  }
+
+  /** Gets the MVC controller that this call is made from, if any. */
+  MicrosoftAspNetCoreMvcController getController() {
+    result = this.getEnclosingCallable().getDeclaringType()
+  }
+
+  /** Gets the name of the MVC controller that this call is made from, if any. */
+  string getControllerName() { result + "Controller" = this.getController().getName() }
+
+  /** Gets the name of the Area that the controller of this call belongs to, if any. */
+  string getAreaName() {
+    exists(Attribute attr |
+      attr = this.getController().getAnAttribute() and
+      attr.getType().hasQualifiedName("Microsoft.AspNetCore.Mvc", "AreaAttribute") and
+      result = attr.getArgument(0).(StringLiteral).getValue()
+    )
+  }
+
+  /** `result` is `true` if this call is from a controller that is an Area, and `false` otherwise. */
+  boolean hasArea() { if exists(this.getAreaName()) then result = true else result = false }
+}
+
+/** A compiler-generated Razor page. */
+class RazorPage extends Class {
+  AssemblyAttribute attr;
+
+  RazorPage() {
+    attr.getFile() = this.getFile() and
+    attr.getType()
+        .hasQualifiedName("Microsoft.AspNetCore.Razor.Hosting", "RazorCompiledItemAttribute")
+  }
+
+  /**
+   * Gets the filepath of the source file that this class was generated from,
+   * relative to the application root.
+   */
+  string getSourceFilepath() { result = attr.getArgument(2).(StringLiteral).getValue() }
+}
+
+private class ViewCallFlowStep extends TaintTracking::AdditionalTaintStep {
```

Comment:
Should this really be a value step and not a taint step?

Reply:
Perhaps. Do we want to add a mechanism for defining additional value steps for all configurations? And are there likely to be many cases where this matters for precision?

Reply:
It matters, especially if data is stored inside a field/property of the model (which I guess is often the case?), since only value-preserving steps are allowed when tracking data stored in a field/property. However, we need to be sure that the step is indeed value-preserving.

We could add a mechanism for adding additional value steps, just like you added one for additional taint steps.